src/Substance.hs

```diff
@@ -4,535 +4,450 @@
+                   substInsert sigma (TypeVarY ftv) (AT (TTypeVar atv))
+substHelper varEnv sigma ((KT (TTypeVar atv)), (KT (TConstr (ConstructorInvoker ftc argsFT pos)))) =
+                   error ("Argument type " ++ (show atv) ++ " doesn't match expected type " ++ (show ftc))
+substHelper varEnv sigma ((KT (TConstr (ConstructorInvoker atc argsAT pos))), (KT (TTypeVar ftv))) =
+                   substInsert sigma (TypeVarY ftv) (AT (TConstr (ConstructorInvoker atc argsAT pos)))
+substHelper varEnv sigma ((KT (TConstr (ConstructorInvoker atc argsAT pos1))), (KT (TConstr (ConstructorInvoker ftc argsFT pos2)))) =
+                   if ((atc `elem` (declaredNames varEnv)) || (ftc `elem` (declaredNames varEnv)))
+                   then substHelper2 varEnv sigma ((AVar (VarConst atc)), (AVar (VarConst ftc)))
+                   else if (atc /= ftc) 
+                        then error ("Argument type " ++ (show atc) ++ " doesn't match expected type " ++ (show ftc))
+                        else let args   = zip argsAT argsFT
+                                 sigma2 = foldl (substHelper2 varEnv) sigma args
+                              in sigma2
+
+substHelper2 :: VarEnv -> M.Map Y Arg -> (Arg, Arg) -> M.Map Y Arg
+substHelper2 varEnv sigma ((AVar av), (AVar fv)) =
+                    substInsert sigma (VarY fv) (AVar av)
+substHelper2 varEnv sigma ((AVar av), (AT ft)) =
+                    error ("Argument type's argument " ++ (show av) ++ " doesn't match expected type's argument " ++ (show ft))
+substHelper2 varEnv sigma ((AT at), (AVar fv)) =
+                    error("Argument type's argument " ++ (show at) ++ " doesn't match expected type's argument " ++ (show fv))
+substHelper2 varEnv sigma ((AT at), (AT ft)) =
+                    substHelper varEnv sigma ((KT at), (KT ft))
+
+substInsert :: M.Map Y Arg -> Y -> Arg -> M.Map Y Arg
+substInsert sigma y arg = case M.lookup y sigma of
+                          Nothing -> M.insert y arg $ sigma
+                          arg'  -> if arg /= (fromJust arg')
+                                   then error "Substitutions inconsistent - no subst can exist"
+                                   else sigma
+
+--argTypeLookupHelper :: VarEnv -> Y -> Either K String
+--argTypeLookupHelper varEnv (TypeVarY t) = case M.lookup t (typeVarMap varEnv) of
+--                                   Nothing -> Right "Argument " ++ (show t) ++ " not in environment\n"
+--                                   tType -> Left tType
+--argTypeLookupHelper varEnv (VarY v) = case M.lookup v (varMap varEnv) of
+--                               Nothing -> Right "Argument " ++ (show v) ++ " not in environment\n"
+--                               vType -> Left vType
+
+
+-- --------------------------------------- Substance Loader --------------------------------
+-- | Load all the Substance objects for visualization at Runtime.hs
+
+data SubObjects = SubObjects { subObjs :: [SubObj] }
```

Comment:
if you're not adding more fields, remove the datatype and just use `[SubObj]` everywhere (and simplify all uses of this type below)

Reply:
In fact might be extended in the really near future....